From a1e109fefeb3ce8ed60dd8b2f1db8cbb02c41548 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 19 Mar 2019 23:41:33 +0200
Subject: [PATCH] [SW-0] preboot: remove device name from code

No more goya/gaudi in code.
This will make project cleanup easier before upstream to
linux-firmware
Any Gaudi code must not be published at this point of time

Change-Id: If122f91a684c7cadaeb40f9b38b9909510e94f7a
---

diff --git a/Makefile b/Makefile
index 19a692b..f3e40a1 100644
--- a/Makefile
+++ b/Makefile
@@ -9,11 +9,7 @@
 CFLAGS += -include ../uboot-custom/include/linux/kconfig.h
 CFLAGS += -Wall -Werror
 CFLAGS += -c
-ifeq ($(HL_DEV),goya)
-	CFLAGS += -DDEV_IS_GOYA
-else
-	CFLAGS += -DDEV_IS_GAUDI
-endif
+CFLAGS += -I ./ -I ./common/
 
 ifeq ($(HL_TARGET),pldm)
 	CFLAGS += -DTARGET_IS_PLDM
@@ -21,12 +17,17 @@
 ifneq ($(TARGET_IS_TESTER),)
 	CFLAGS += -DTARGET_IS_TESTER
 endif
-LDFLAGS := -g -T preboot.lds -nostdlib -static
+
+SFLAGS := $(CFLAGS)
+CFLAGS += -include $(HL_DEV)/$(HL_DEV).h
+
+LDFLAGS := -g -T common/preboot.lds -nostdlib -static
 LDFLAGS += -L ../../host/opt/ext-toolchain/lib/gcc/aarch64-linux-gnu/6.4.1
 LDLIBS := -lgcc
 
-PREBOOT_SRCS_C := $(shell ls *.c)
-PREBOOT_SRCS_S += $(shell ls *.S)
+PREBOOT_SRCS_C := $(shell ls common/*.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/*.c)
+PREBOOT_SRCS_S := $(shell ls common/*.S)
 
 PREBOOT_OBJECTS := $(PREBOOT_SRCS_C:.c=.o)
 PREBOOT_OBJECTS += $(PREBOOT_SRCS_S:.S=.o)
@@ -49,7 +50,7 @@
 	$(CC) $(CFLAGS) -o $@ $<
 
 %.o: %.S $(AUTO_GENERATED_FILE)
-	$(CC) $(CFLAGS) -o $@ $<
+	$(CC) $(SFLAGS) -o $@ $<
 
 clean:
 	rm -f $(PREBOOT_OBJECTS) $(AUTO_GENERATED_FILE)
diff --git a/common/preboot.c b/common/preboot.c
new file mode 100644
index 0000000..861afeb
--- /dev/null
+++ b/common/preboot.c
@@ -0,0 +1,154 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <common.h>
+#include <linux/io.h>
+#include <mach/hl_regs.h>
+#include <mach/hl_fw_if.h>
+#include "auto_generated.h"
+#include <stddef.h>
+#include <stdbool.h>
+#include "uart.h"
+#include "spi.h"
+#include "pci.h"
+
+#define VERSION_MAX_LEN		128
+
+#define PREBOOT_VER_STR	"Preboot version " PREBOOT_VER \
+	" (" PREBOOT_DATE " - " PREBOOT_TIME ")"
+
+/*
+ * Using scratchpad registers for
+ * passing flash layout to CPU SW
+ */
+static void preboot_copy_prepare(unsigned int *spl_off)
+{
+	unsigned int spl_offset = 0xa0000;
+	unsigned int uboot_offset = 0xb0000;
+	unsigned int modifiable_image_pending;
+	unsigned int image_base_offset;
+	unsigned int next_image, next_image_stable, next_image_modifiable;
+	bool try_new_image = false;
+	char *spl_image = "candidate";
+
+	modifiable_image_pending = readl(mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
+	if (modifiable_image_pending == FLOPS_UPD_PENDING) {
+		next_image_modifiable =
+			(*(unsigned int *)(HL_SPI_BASE + 0x60004) & 0xff) - '1';
+		if (next_image_modifiable < 2) {
+			next_image = next_image_modifiable;
+			writel((next_image + FLOPS_UPD_NXT_IMG1),
+					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
+			try_new_image = true;
+		} else
+			writel(FLOPS_UPD_NOOP,
+					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
+	}
+
+	if (!try_new_image) {
+		spl_image = "stable";
+		next_image_stable =
+			(*(unsigned int *)(HL_SPI_BASE + 0x60000) & 0xff) - '1';
+		if (next_image_stable < 2)
+			next_image = next_image_stable;
+		else
+			next_image = 0;
+	}
+
+	if (!next_image)
+		image_base_offset = 0;
+	else
+		image_base_offset = 0x800000;
+
+	preboot_puts_raw("Preparing to run SPL ");
+	preboot_puts_raw(spl_image);
+	preboot_puts_raw(" (offset: 0x");
+	preboot_puth_raw(image_base_offset, false);
+	preboot_puts(")\n");
+
+	*spl_off = image_base_offset + spl_offset;
+	writel((image_base_offset + uboot_offset)
+			, mmPSOC_GLOBAL_CONF_SCRATCHPAD_30);
+}
+
+
+static void preboot_copy_spl(void)
+{
+	unsigned int  spl_offset;
+
+	preboot_copy_prepare(&spl_offset);
+
+	spi_copy_spl(spl_offset);
+}
+
+static void preboot_pcie_init(void)
+{
+	if (readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS) & SKIP_PCIE_INIT) {
+		preboot_puts("PCIe initialization is skipped!\n");
+		return;
+	}
+
+	preboot_pcie_wait_for_link();
+	preboot_puts("PCIe link is UP!\n");
+}
+
+#ifndef TARGET_IS_TESTER
+static const char *preboot_ver_str = PREBOOT_VER_STR;
+static char preboot_btl_ver_str[VERSION_MAX_LEN] = "BTL version ";
+
+static void preboot_version(void)
+{
+	u64 offset = (u64) preboot_btl_ver_str;
+	u32 btlver = readl(mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
+	int i, btl_ver_sz;
+
+	preboot_puts("\n");
+	preboot_puts_raw(preboot_btl_ver_str);
+	preboot_puth(btlver, true);
+	preboot_puts_raw(preboot_ver_str);
+	preboot_puts("\n");
+
+	/* get to the end of the predefined btl string */
+	for (i = 0; i < VERSION_MAX_LEN; i++)
+		if (preboot_btl_ver_str[i] == '\0')
+			break;
+
+	/* convert and copy the btl version */
+	i += preboot_uint2hexstring(&preboot_btl_ver_str[i], btlver, true);
+
+	/* separate the btl from preboot */
+	preboot_btl_ver_str[i++] = '\n';
+	btl_ver_sz = i;
+
+	/* copy the preboot version */
+	for (int j = 0; j < VERSION_MAX_LEN - btl_ver_sz - 1; j++) {
+		preboot_btl_ver_str[i++] = preboot_ver_str[j];
+		if (preboot_ver_str[j] == '\0')
+			break;
+	}
+
+	writel((u32)(offset & ~CONFIG_SYS_SDRAM_BASE),
+			mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
+}
+#else	/* TARGET_IS_TESTER */
+static inline void preboot_version(void) {}
+#endif	/* TARGET_IS_TESTER */
+
+void preboot_main(void)
+{
+	void (*spl_start)(void) = (void *)CONFIG_SPL_TEXT_BASE;
+
+	writel(CPU_BOOT_STATUS_IN_PREBOOT, mmPSOC_GLOBAL_CONF_WARM_REBOOT);
+
+	preboot_version();
+	preboot_pcie_init();
+	preboot_delay();
+	hl_pll_init();
+	preboot_copy_spl();
+
+	spl_start();
+}
diff --git a/common/preboot.lds b/common/preboot.lds
new file mode 100644
index 0000000..ea1f531
--- /dev/null
+++ b/common/preboot.lds
@@ -0,0 +1,23 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+MEMORY { .sram : ORIGIN = 0x7ff0000000,
+		LENGTH = 0x10000 }
+
+OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
+OUTPUT_ARCH(aarch64)
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(8);
+		common/start.o (.text*)
+		* (.text*)
+	} >.sram
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
diff --git a/common/spi.c b/common/spi.c
new file mode 100644
index 0000000..1315bd2
--- /dev/null
+++ b/common/spi.c
@@ -0,0 +1,227 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <config.h>
+#include <linux/io.h>
+#include <linux/sizes.h>
+#include <linux/bitops.h>
+#include <mach/hl_regs.h>
+
+#include "uart.h"
+
+#define mmPSOC_SPI_CTRLR0		(HL_CFG_BASE + 0xC43000)
+#define mmPSOC_SPI_SSIENR		(HL_CFG_BASE + 0xC43008)
+#define mmPSOC_SPI_SER			(HL_CFG_BASE + 0xC43010)
+#define mmPSOC_SPI_BAUDR		(HL_CFG_BASE + 0xC43014)
+#define mmPSOC_SPI_TXFTLR		(HL_CFG_BASE + 0xC43018)
+#define mmPSOC_SPI_RXFTLR		(HL_CFG_BASE + 0xC4301C)
+#define mmPSOC_SPI_TXFLR		(HL_CFG_BASE + 0xC43020)
+#define mmPSOC_SPI_RXFLR		(HL_CFG_BASE + 0xC43024)
+#define mmPSOC_SPI_IMR			(HL_CFG_BASE + 0xC4302C)
+#define mmPSOC_SPI_DR0			(HL_CFG_BASE + 0xC43060)
+#define SPL_SIZE			SZ_64K
+#define CMD_WRITE_STATUS		0x01
+#define CMD_READ			0x03
+#define CMD_WRITE_DISABLE		0x04
+#define CMD_READ_STATUS			0x05
+#define CMD_WRITE_ENABLE		0x06
+#define CMD_READ_CONF			0x15
+#define READ_STATUS_WIP			BIT(0)
+#define READ_STATUS_BP0			BIT(2)
+#define READ_STATUS_BP1			BIT(3)
+#define READ_STATUS_BP2			BIT(4)
+#define READ_STATUS_BP3			BIT(5)
+#define READ_STATUS_SRWD		BIT(7)
+#define READ_STATUS_BP0_3		(READ_STATUS_BP0 | READ_STATUS_BP1 | \
+					READ_STATUS_BP2 | READ_STATUS_BP3)
+#define PROT_MASK			(READ_STATUS_BP0_3 | READ_STATUS_SRWD)
+#define frame_size_bits(b)		((b) << 16)
+
+/*
+ * SSI device Chip Select (CS0) is muxed with GPIO53
+ * Using this we can realy do SSI bus claim/release
+ * GPIO53 is set through mmPSOC_GPIO3 where each
+ * mmPSOC_GPIO serve 16 GPIOs
+ * Note: CS0 is active low
+ */
+static void spi_ssi_cs0_init(void)
+{
+	unsigned int reg, mask = BIT_MASK(53 % 16);
+
+	reg  = readl(mmPSOC_GPIO3_DATAOUT);
+	reg  &= ~mask;
+	writel(reg,  mmPSOC_GPIO3_DATAOUT);
+
+	reg = readl(mmPSOC_GPIO3_OUTENSET);
+	reg |=  mask;
+	writel(reg, mmPSOC_GPIO3_OUTENSET);
+}
+
+static void spi_claim_bus(void)
+{
+	writel(1, mmPSOC_GLOBAL_CONF_PAD_SEL_53);
+}
+
+static void spi_release_bus(void)
+{
+	writel(0, mmPSOC_GLOBAL_CONF_PAD_SEL_53);
+}
+
+static inline void spi_dev_update_frame_size(unsigned int len)
+{
+	writel(0, mmPSOC_SPI_SSIENR);
+	writel(frame_size_bits((len * 8) - 1), mmPSOC_SPI_CTRLR0);
+	writel(1, mmPSOC_SPI_SSIENR);
+}
+
+static inline int spi_fifo_is_empty(unsigned long level_reg)
+{
+	return !readl(level_reg);
+}
+
+static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
+{
+	unsigned int tmp;
+
+	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
+		;
+	writel(cmd, mmPSOC_SPI_DR0);
+
+	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
+		;
+	tmp = readl(mmPSOC_SPI_DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+static void spi_special_cmd(unsigned int cmd, unsigned char *val)
+{
+	unsigned char tmp;
+
+	spi_dev_update_frame_size(sizeof(tmp));
+
+	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
+		;
+	writeb(cmd, mmPSOC_SPI_DR0);
+
+	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
+		;
+	tmp = readb(mmPSOC_SPI_DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+/* Hold chip select (CS) during transaction */
+static void spi_char_cmd_csel(unsigned char cmd, unsigned char *val)
+{
+	spi_claim_bus();
+
+	spi_special_cmd(cmd, NULL);
+	if (val)
+		spi_special_cmd(0, val);
+
+	spi_release_bus();
+}
+
+static void spi_special_cmd_csel(unsigned int cmd, unsigned char *val,
+								size_t size)
+{
+	int i;
+
+	spi_claim_bus();
+
+	spi_dev_update_frame_size(sizeof(*val));
+
+	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
+		;
+	writeb(cmd, mmPSOC_SPI_DR0);
+	for (i = 0; i < size; i++) {
+		while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
+			;
+		writeb(val[i], mmPSOC_SPI_DR0);
+	}
+
+	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
+		;
+	readb(mmPSOC_SPI_DR0);
+	for (i = 0; i < size; i++) {
+		while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
+			;
+		readb(mmPSOC_SPI_DR0);
+	}
+
+	spi_release_bus();
+}
+
+static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
+{
+	unsigned char sr;
+
+	spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
+	spi_special_cmd_csel(cmd, buf, size);
+	do
+		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	while (sr & READ_STATUS_WIP);
+	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
+}
+
+static unsigned char spi_prot(void)
+{
+	unsigned char sr, cr, data[2];
+
+	spi_char_cmd_csel(CMD_READ_CONF, &cr);
+	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	sr |= PROT_MASK;
+	data[0] = sr;
+	data[1] = cr;
+	spi_common_write(CMD_WRITE_STATUS, data, sizeof(data));
+	return sr;
+}
+
+static void spi_read_spl(unsigned int offset)
+{
+	void *dst = (void *)CONFIG_SPL_TEXT_BASE;
+	unsigned int cmd = (CMD_READ << 24) | offset;
+	int i;
+
+	spi_claim_bus();
+	spi_dev_update_frame_size(sizeof(cmd));
+	spi_dev_dword_cmd(cmd, NULL);
+	for (i = 0; i < SPL_SIZE; i += sizeof(unsigned int))
+		spi_dev_dword_cmd(0, dst + i);
+	spi_dev_dword_cmd(0, NULL);
+	spi_release_bus();
+}
+
+static void spi_init(void)
+{
+	spi_ssi_cs0_init();
+	writel(0x00000000, mmPSOC_SPI_SSIENR);
+	writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
+	writel(0x00000000, mmPSOC_SPI_RXFTLR);
+	writel(0x00000000, mmPSOC_SPI_TXFTLR);
+	writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
+	writel(0x00000001, mmPSOC_SPI_SER);
+	writel(0x00000001, mmPSOC_SPI_SSIENR);
+}
+
+void spi_copy_spl(unsigned int spl_offset)
+{
+	unsigned char sr;
+
+	spi_init();
+	spi_read_spl(spl_offset);
+	sr = spi_prot();
+	preboot_puts("SPL copying from flash is completed\n");
+	if ((sr & PROT_MASK) == PROT_MASK)
+		preboot_puts("Flash may be protected, depend on WP# signal\n");
+	else
+		preboot_puts("Flash is not protected\n");
+
+}
diff --git a/common/spi.h b/common/spi.h
new file mode 100644
index 0000000..5d2051c
--- /dev/null
+++ b/common/spi.h
@@ -0,0 +1,14 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef _HL_SPI_H_
+#define _HL_SPI_H_
+
+
+void spi_copy_spl(unsigned int spl_offset);
+
+#endif /* _HL_SPI_H_ */
diff --git a/common/start.S b/common/start.S
new file mode 100644
index 0000000..f988923
--- /dev/null
+++ b/common/start.S
@@ -0,0 +1,18 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+/*
+ * This is preboot entry point
+ * Setting sp/fp/lr and jumping to C code
+ */
+.globl	_start
+_start:
+
+	ldr	x0, =(_start + 0x10000)
+	bic	sp, x0, #0xf	/* 16-byte alignment for ABI compliance */
+	mov	x29, sp		/* Set FP (x29) */
+	bl	preboot_main	/* Set LR (x30) */
diff --git a/common/uart.c b/common/uart.c
new file mode 100644
index 0000000..eaba0c6
--- /dev/null
+++ b/common/uart.c
@@ -0,0 +1,113 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <linux/io.h>
+#include <mach/hl_regs.h>
+
+#define mmPSOC_UART_0_DATA		(HL_CFG_BASE + 0xC45000)
+#define mmPSOC_UART_0_STATUS		(HL_CFG_BASE + 0xC45004)
+#define UART_SR_TXFULL			(1 << 0) /* TX Buffer Full */
+#define STR_UINT_SZ			8
+
+void preboot_puts_raw(const char *str)
+{
+	int i;
+
+	for (i = 0; str[i] && (i < 128); i++) {
+		while (readl(mmPSOC_UART_0_STATUS) & UART_SR_TXFULL)
+			;
+		writel(str[i], mmPSOC_UART_0_DATA);
+		preboot_putc_virt_uart(str[i]);
+	}
+}
+
+void preboot_puts(const char *str)
+{
+	preboot_puts_raw(str);
+	preboot_puts_raw("\r");
+}
+
+/*
+ * preboot_byte2char	- convert half byte to char.
+ *
+ * This function takes a byte number and converts its value to
+ * a character representing the value (e.g. 0x8c can be converted to
+ * '8' or 'c').
+ */
+static inline char preboot_byte2char(u8 byte, bool upper)
+{
+	/* upper or lower 4 bits of the byte */
+	byte = upper ? byte >> 4 : byte & 0xf;
+
+	/*
+	 * convert: if half byte is a number [0..9] => use '0' as an offset
+	 *	    if half byte is a letter [0xa..0xf] => use 'a' as a base,
+	 *	    but since 'a' is not 0, we need to subtract
+	 *	    its base offset - 0xa.
+	 */
+	return byte + ((byte >= 0xa) ? 'a' - 0xa : '0');
+}
+
+/*
+ * preboot_uint2hexstring	- convert uint to a hex values string.
+ *
+ * Take a number and provide its hexadecimal string representation.
+ */
+int preboot_uint2hexstring(char *str, unsigned int num, bool pad)
+{
+	int j = 0;
+	char tmp;
+
+	/*
+	 * Here we walk on each byte of uint num in the reverse order, so
+	 * we build the string in the correct order.
+	 */
+	for (int i = 3; i >= 0; i--) {
+		u8 byte = (u8)((num >> (i * 8)) & 0xff);
+		/* here we run the loop on 2 parts of byte: upper and lower */
+		for (int k = 0; k < 2; k++) {
+			tmp = preboot_byte2char(byte, !k);
+			if (tmp == '0')	/* care of the leading zeroes (pad) */
+				pad = true;
+			if (pad)
+				str[j++] = tmp;
+		}
+	}
+
+	/* for !pad, return how many characters were actually written */
+	return j;
+}
+
+void preboot_puth_raw(unsigned int val, bool pad)
+{
+	char str[STR_UINT_SZ + 1];
+	int val_sz;
+
+	val_sz = preboot_uint2hexstring(str, val, pad);
+	val_sz = val_sz > STR_UINT_SZ ? STR_UINT_SZ : val_sz;
+	str[val_sz] = '\0';
+	preboot_puts_raw(str);
+}
+
+void preboot_puth(unsigned int val, bool pad)
+{
+	preboot_puth_raw(val, pad);
+	preboot_puts("\n");
+}
+
+static inline void preboot_puts_hex(const char *log, unsigned int val, bool pad)
+{
+	preboot_puts_raw(log);
+	preboot_puts_raw("0x");
+	preboot_puth(val, pad);
+}
+
+static inline void preboot_puts_char(const char *str, unsigned char val)
+{
+	preboot_puts_hex(str, val, false);
+}
+
diff --git a/common/uart.h b/common/uart.h
new file mode 100644
index 0000000..0d298f6
--- /dev/null
+++ b/common/uart.h
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef _HL_UART_H_
+#define _HL_UART_H_
+
+
+void preboot_puts_raw(const char *str);
+void preboot_puts(const char *str);
+int preboot_uint2hexstring(char *str, unsigned int num, bool pad);
+void preboot_puth_raw(unsigned int val, bool pad);
+void preboot_puth(unsigned int val, bool pad);
+
+#endif /* _HL_UART_H_ */
diff --git a/gaudi/gaudi.c b/gaudi/gaudi.c
new file mode 100644
index 0000000..de6092a
--- /dev/null
+++ b/gaudi/gaudi.c
@@ -0,0 +1,21 @@
+/*
+ * (C) Copyright 2019 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <linux/io.h>
+#include <mach/hl_regs.h>
+
+void preboot_pcie_wait_for_link(void)
+{
+	while (1) {
+		unsigned int smlh, rdlh;
+
+		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
+		if (smlh && rdlh)
+			break;
+	}
+}
diff --git a/gaudi/gaudi.h b/gaudi/gaudi.h
new file mode 100644
index 0000000..4f7ca9e
--- /dev/null
+++ b/gaudi/gaudi.h
@@ -0,0 +1,15 @@
+/*
+ * (C) Copyright 2019 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+/* Use DumpSel[8] to skip PCIe initialization */
+#define SKIP_PCIE_INIT BIT(16)
+
+void preboot_pcie_wait_for_link(void);
+static inline void preboot_delay(void) {}
+static inline void preboot_putc_virt_uart(const char c) {}
+static inline void hl_pll_init(void) {}
+
diff --git a/goya/goya.c b/goya/goya.c
new file mode 100644
index 0000000..7ea9ce2
--- /dev/null
+++ b/goya/goya.c
@@ -0,0 +1,150 @@
+/*
+ * (C) Copyright 2019 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <common.h>
+#include <linux/io.h>
+#include <mach/hl_regs.h>
+
+#define BTL_COUNTER_MAX		BIT(22)
+#define SYS_REF_CLOCK		50000000
+
+static int preboot_wait_for_bit(unsigned long reg, unsigned long timeout_ns)
+{
+	unsigned int val;
+	unsigned long count = (timeout_ns * SYS_REF_CLOCK) / 1000000000UL;
+
+	while (count--) {
+		val = readl(reg);
+		if (val & 1)
+			return 0;
+	}
+
+	return 1;
+}
+
+static void preboot_wait_for_ext_ld_done(void)
+{
+	/*
+	 * Wait for HW FSM to complete 16KB copy to PHY
+	 * this is enough to set ext_ld_done.
+	 * FSM will set ext_ld_done after 32KB but why wait
+	 * if we can already move on
+	 */
+	while ((readl(mmPCIE_PHY_CR_PARA_CMD) & 0xFFFF) <= 0xD780)
+		;
+	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0);
+	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1);
+	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2);
+	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3);
+}
+
+static void preboot_wait_for_stable_pclk(void)
+{
+	int ret;
+
+	while (1) {
+		ret = 0;
+		/* Wait for PCLK stable - check MPLLA [no timeout] */
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_0, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_1, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_2, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_3, 100);
+		if (!ret)
+			break;
+
+		ret = 0;
+		/* Wait for PCLK stable - check MPLLB [no timeout] */
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_0, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_1, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_2, 100);
+		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_3, 100);
+		if (!ret)
+			break;
+	}
+}
+
+/* Handle PCIe erratum: make sure reset_n de-asserted only after perst */
+void preboot_pcie_wait_for_link(void)
+{
+	preboot_wait_for_ext_ld_done();
+
+	/* Wait for PRSTn */
+	while (!readl(mmPSOC_GLOBAL_CONF_PRSTN))
+		;
+
+	/* Enable FW FSM after PRSTn */
+	writel(1, mmPSOC_GLOBAL_CONF_PCI_FW_FSM);
+
+	preboot_wait_for_stable_pclk();
+
+	while (1) {
+		unsigned int prstn, ltssm, smlh, rdlh, hot_reset;
+
+		prstn = readl(mmPSOC_GLOBAL_CONF_PRSTN);
+		ltssm = readl(mmPCIE_AUX_LTSSM_EN);
+		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
+		hot_reset = readl(mmPCIE_CORE_HOT_RST);
+
+		if ((hot_reset & BIT(0)) == BIT(0)) {
+			writel(0x10, mmPCIE_CORE_HOT_RST);
+			writel(1, mmPCIE_AUX_LTSSM_EN);
+		}
+
+		if (prstn && smlh && rdlh)
+			break;
+
+		/*
+		 * Some host assert prstn during link up while moving to GEN3
+		 * This cause BAR size to be reset
+		 * We stop link up process by disabling ltssm
+		 * till BAR is reconfigured
+		 */
+		if (!prstn && ltssm) {
+			writel(0, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+
+		if (prstn && !ltssm) {
+			writel(1, mmPCIE_AUX_LTSSM_EN);
+			continue;
+		}
+	}
+}
+
+/* Handle BTL erratum: exhaust timeout counter before any SPI access */
+void preboot_delay(void)
+{
+	unsigned long cntpct, end;
+	int sysclk = SYS_REF_CLOCK, i;
+
+	/* timer init */
+	writel(sysclk, mmCPU_TIMESTAMP_CNTFID0);
+	writel(0x3, mmCPU_TIMESTAMP_CNTCR); /* EN | HDBG */
+
+	/* get begin counter value and calc end counter value */
+	asm volatile("mrs %0, cntpct_el0" : "=r" (cntpct));
+	end = cntpct + BTL_COUNTER_MAX;
+
+	for (i = 0; i < BTL_COUNTER_MAX; i++) {
+		asm volatile("mrs %0, cntpct_el0" : "=r" (cntpct));
+		if (cntpct >= end)
+			break;
+	}
+}
+
+void preboot_putc_virt_uart(const char c)
+{
+	/*
+	 * update vUART to HOST
+	 * DATA: mmPCIE_DB_RSV_RSV_0
+	 * STATUS: mmPCIE_DB_RSV_RSV_1
+	 */
+	writel(c, mmPCIE_DB_RSV_RSV_0);
+	writel(1, mmPCIE_DB_RSV_RSV_1);
+}
+
diff --git a/goya/goya.h b/goya/goya.h
new file mode 100644
index 0000000..a0147b0
--- /dev/null
+++ b/goya/goya.h
@@ -0,0 +1,14 @@
+/*
+ * (C) Copyright 2019 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+/* Use DumpSel[8] to skip PCIe initialization */
+#define SKIP_PCIE_INIT BIT(21)
+
+void preboot_pcie_wait_for_link(void);
+void preboot_delay(void);
+void preboot_putc_virt_uart(const char c);
+void hl_pll_init(void);
diff --git a/goya/pll.c b/goya/pll.c
new file mode 100644
index 0000000..c30e1f5
--- /dev/null
+++ b/goya/pll.c
@@ -0,0 +1,246 @@
+#include <common.h>
+#include <command.h>
+#include <linux/delay.h>
+#include <linux/io.h>
+#include <linux/types.h>
+#include "uart.h"
+
+#define HL_PLL_NR		0x100
+#define HL_PLL_NF		0x104
+#define HL_PLL_OD		0x108
+#define HL_PLL_NB		0x10c
+#define HL_PLL_CFG		0x110
+#define HL_PLL_CFG_LOCK		(1 << 4)
+#define HL_PLL_LOCK_INTR	0x128
+#define HL_PLL_DATA_CHNG	0x130
+#define HL_PLL_RST		0x134
+#define HL_PLL_DIV_FACTOR	0x200
+#define HL_PLL_DIV_FACTOR_CMD	0x220
+#define HL_PLL_DIV_SEL		0x280
+#define HL_PLL_DIV_EN		0x2a0
+
+#define HL_PLL_MAX_WAIT_TIME	20	/* 20 us */
+#define HL_PLL_LOCK_TOUT	5000	/* 5 ms */
+
+#define USEC_PER_SEC		1000000
+#define CPU_CYCELS_PER_USEC	(1200000000 / USEC_PER_SEC)
+#define BYPASS_CYCLES_PER_USEC	(50000000 / USEC_PER_SEC)
+#define MIN_CYCLES_PER_ITER	5
+
+#define mmCPU_PLL_BASE		0x7FFC4A2000ull
+#define mmIC_PLL_BASE		0x7FFC4A3000ull
+#define mmPSOC_MME_PLL_BASE	0x7FFCC71000ull
+
+enum hl_pll_bypass_val {
+	HL_PLL_REF_CLK,
+	HL_PLL_PLL_CLK,
+	HL_PLL_DIV_REF,
+	HL_PLL_DIV_PLL,
+};
+
+enum hl_pll_div_out {
+	HL_PLL_DIV0,
+	HL_PLL_DIV1,
+	HL_PLL_DIV2,
+	HL_PLL_DIV3,
+	HL_PLL_DIV_SIZE,
+};
+
+struct hl_pll_cfg {
+	char	*name;
+	void	*base;
+	u32	nr;
+	u32	nf;
+	u32	od;
+	u32	nb;
+	u32	div_factor[HL_PLL_DIV_SIZE];
+};
+
+static struct hl_pll_cfg goya_pll_cfg[] = {
+	{.name = "CPU", .base = (void *)mmCPU_PLL_BASE,
+	 .nr = 0, .nf = 47, .od = 1, .nb = 23, .div_factor = {0, 23, 1, -1} },
+	{.name = "IC", .base = (void *)mmIC_PLL_BASE,
+	 .nr = 0, .nf = 63, .od = 15, .nb = 31, .div_factor = {0, 1, 1, 2} },
+	{.name = "MME", .base = (void *)mmPSOC_MME_PLL_BASE,
+	 .nr = 0, .nf = 63, .od = 15, .nb = 31, .div_factor = {0, 1, 2, 8} },
+};
+
+static int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
+{
+	int pll_num = ARRAY_SIZE(goya_pll_cfg);
+
+	*hl_pll_cfg = goya_pll_cfg;
+
+	return pll_num;
+}
+
+static enum hl_pll_bypass_val hl_pll_bypass_get(struct hl_pll_cfg *cfg,
+							enum hl_pll_div_out out)
+{
+	int offset = out * sizeof(u32);
+
+	return readl(cfg->base + HL_PLL_DIV_SEL + offset);
+}
+
+static void hl_pll_udelay(unsigned int usec)
+{
+	long cnt;
+	/*
+	 * Assumptions:
+	 * 1) the CPU PLL offset in the array is 0
+	 * 2) DIV1 is the PLL output to timestamp counter.
+	 */
+	enum hl_pll_bypass_val bval =
+			hl_pll_bypass_get(goya_pll_cfg, HL_PLL_DIV1);
+
+	if (bval == HL_PLL_REF_CLK || bval == HL_PLL_DIV_REF)
+		cnt = BYPASS_CYCLES_PER_USEC;
+	else
+		cnt = CPU_CYCELS_PER_USEC;
+
+	cnt = (cnt * usec) / MIN_CYCLES_PER_ITER;
+
+	while (cnt--)
+		;
+}
+
+static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
+{
+	writel(reset, cfg->base + HL_PLL_RST);
+}
+
+static inline int hl_pll_is_locked(struct hl_pll_cfg *cfg)
+{
+	return readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
+}
+
+static inline void hl_pll_clear_lock_status(struct hl_pll_cfg *cfg)
+{
+	u32 reg = readl(cfg->base + HL_PLL_CFG);
+
+	writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
+}
+
+static int hl_pll_cfg(struct hl_pll_cfg *cfg)
+{
+	int tout = HL_PLL_LOCK_TOUT;
+
+	hl_pll_reset(cfg, true);
+
+	writel(cfg->nr, cfg->base + HL_PLL_NR);
+	writel(cfg->nf, cfg->base + HL_PLL_NF);
+	writel(cfg->od, cfg->base + HL_PLL_OD);
+	writel(cfg->nb, cfg->base + HL_PLL_NB);
+	writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
+
+	/* after updating the coefficients we need to wait >=5us */
+	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
+
+	hl_pll_reset(cfg, false);
+
+	/* Wait >500 (10us) reference cycles until PLL re-locks. */
+	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
+
+	/* to be on a safe side, poll on the PLL lock status bit */
+	while (!hl_pll_is_locked(cfg) && tout--)
+		hl_pll_udelay(1);
+
+	if (!tout) {
+		preboot_puts_raw("PLL: Timeout waiting for ");
+		preboot_puts_raw(cfg->name);
+		preboot_puts_raw(" PLL lock...\n\r");
+		return 1;
+	}
+
+	/* w/a the h/w erratum: lock status bit isn't getting cleared */
+	hl_pll_clear_lock_status(cfg);
+
+	return 0;
+}
+
+static void hl_pll_bypass(struct hl_pll_cfg *cfg, bool bypass)
+{
+	int sel;
+	void *reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		if (cfg->div_factor[sel] == -1)
+			continue;
+
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+
+		if (bypass)
+			writel(HL_PLL_REF_CLK, reg);
+		else
+			writel(cfg->div_factor[sel] ?
+					HL_PLL_DIV_PLL : HL_PLL_PLL_CLK, reg);
+	}
+
+	/* just to be on a safe side, wait for bypass to propagate */
+	hl_pll_udelay(1);
+}
+
+static void hl_pll_div_init(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	void *div_base;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		if (cfg->div_factor[sel] == -1)
+			continue;
+
+		div_base = cfg->base + (sel * sizeof(u32));
+
+		writel(cfg->div_factor[sel], div_base + HL_PLL_DIV_FACTOR);
+		writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
+		writel(1, div_base + HL_PLL_DIV_EN);
+	}
+
+	/* just to be on a safe side, wait for DIV init to propagate */
+	hl_pll_udelay(1);
+}
+
+static void hl_pll_ack_int(struct hl_pll_cfg *cfg)
+{
+	writel(1, cfg->base + HL_PLL_LOCK_INTR);
+}
+
+static int hl_pll_set(struct hl_pll_cfg *cfg)
+{
+	int ret;
+
+	hl_pll_bypass(cfg, true);
+
+	ret = hl_pll_cfg(cfg);
+
+	hl_pll_div_init(cfg);
+
+	hl_pll_bypass(cfg, false);
+
+	hl_pll_ack_int(cfg);
+
+	return ret;
+}
+
+/**
+ * hl_pll_init - Initialize PLL registers
+ */
+void hl_pll_init(void)
+{
+	int i, pll_num;
+	struct hl_pll_cfg *hl_pll_cfg;
+	enum hl_pll_bypass_val bval;
+
+	pll_num = hl_pll_get_cfg(&hl_pll_cfg);
+
+	for (i = 0; i < pll_num; i++) {
+		/* DIV0 is the undivided PLL output */
+		bval = hl_pll_bypass_get(hl_pll_cfg + i, HL_PLL_DIV0);
+		if ((u64)hl_pll_cfg[i].base == mmCPU_PLL_BASE &&
+				bval != HL_PLL_REF_CLK &&
+				bval != HL_PLL_DIV_REF)
+			continue;
+
+		hl_pll_set(hl_pll_cfg + i);
+	}
+}
diff --git a/pci.c b/pci.c
deleted file mode 100644
index 670c2ab..0000000
--- a/pci.c
+++ /dev/null
@@ -1,149 +0,0 @@
-/*
- * (C) Copyright 2019 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <common.h>
-#include <linux/io.h>
-#include <mach/hl_regs.h>
-#include <mach/hl_fw_if.h>
-#include "preboot.h"
-#include "uart.h"
-
-#ifdef DEV_IS_GOYA
-/* Use DumpSel[8] to skip PCIe initialization */
-#define SKIP_PCIE_INIT BIT(21)
-#else
-#define SKIP_PCIE_INIT BIT(16)
-#endif
-
-#if defined(DEV_IS_GOYA)
-static int preboot_wait_for_bit(unsigned long reg, unsigned long timeout_ns)
-{
-	unsigned int val;
-	unsigned long count = (timeout_ns * SYS_REF_CLOCK) / 1000000000UL;
-
-	while (count--) {
-		val = readl(reg);
-		if (val & 1)
-			return 0;
-	}
-
-	return 1;
-}
-
-static void preboot_wait_for_ext_ld_done(void)
-{
-	/*
-	 * Wait for HW FSM to complete 16KB copy to PHY
-	 * this is enough to set ext_ld_done.
-	 * FSM will set ext_ld_done after 32KB but why wait
-	 * if we can already move on
-	 */
-	while ((readl(mmPCIE_PHY_CR_PARA_CMD) & 0xFFFF) <= 0xD780)
-		;
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3);
-}
-
-static void preboot_wait_for_stable_pclk(void)
-{
-	int ret;
-
-	while (1) {
-		ret = 0;
-		/* Wait for PCLK stable - check MPLLA [no timeout] */
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_0, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_1, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_2, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLA_STATE_3, 100);
-		if (!ret)
-			break;
-
-		ret = 0;
-		/* Wait for PCLK stable - check MPLLB [no timeout] */
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_0, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_1, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_2, 100);
-		ret |= preboot_wait_for_bit(mmPCIE_PHY_PHYX_MPLLB_STATE_3, 100);
-		if (!ret)
-			break;
-	}
-}
-
-/* Handle PCIe errata: make sure reset_n de-asserted only after perst */
-static void preboot_pcie_wait_for_link(void)
-{
-	preboot_wait_for_ext_ld_done();
-
-	/* Wait for PRSTn */
-	while (!readl(mmPSOC_GLOBAL_CONF_PRSTN))
-		;
-
-	/* Enable FW FSM after PRSTn */
-	writel(1, mmPSOC_GLOBAL_CONF_PCI_FW_FSM);
-
-	preboot_wait_for_stable_pclk();
-
-	while (1) {
-		unsigned int prstn, ltssm, smlh, rdlh, hot_reset;
-
-		prstn = readl(mmPSOC_GLOBAL_CONF_PRSTN);
-		ltssm = readl(mmPCIE_AUX_LTSSM_EN);
-		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
-		hot_reset = readl(mmPCIE_CORE_HOT_RST);
-
-		if ((hot_reset & BIT(0)) == BIT(0)) {
-			writel(0x10, mmPCIE_CORE_HOT_RST);
-			writel(1, mmPCIE_AUX_LTSSM_EN);
-		}
-
-		if (prstn && smlh && rdlh)
-			break;
-
-		/*
-		 * Some host assert prstn during link up while moving to GEN3
-		 * This cause BAR size to be reset
-		 * We stop link up process by disabling ltssm
-		 * till BAR is reconfigured
-		 */
-		if (!prstn && ltssm) {
-			writel(0, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-
-		if (prstn && !ltssm) {
-			writel(1, mmPCIE_AUX_LTSSM_EN);
-			continue;
-		}
-	}
-}
-#else
-static void preboot_pcie_wait_for_link(void)
-{
-	while (1) {
-		unsigned int smlh, rdlh;
-
-		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
-		if (smlh && rdlh)
-			break;
-	}
-}
-#endif
-
-void preboot_pcie_init(void)
-{
-	if (readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS) & SKIP_PCIE_INIT) {
-		preboot_puts("PCIe initialization is skipped!\n");
-		return;
-	}
-
-	preboot_pcie_wait_for_link();
-	preboot_puts("PCIe link is UP!\n");
-}
diff --git a/pci.h b/pci.h
deleted file mode 100644
index dfa1b7b..0000000
--- a/pci.h
+++ /dev/null
@@ -1,14 +0,0 @@
-/*
- * Copyright (C) 2019 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_PCI_H_
-#define _HL_PCI_H_
-
-
-void preboot_pcie_init(void);
-
-#endif /* _HL_PCI_H_ */
diff --git a/pll.c b/pll.c
deleted file mode 100644
index 18ee3fc..0000000
--- a/pll.c
+++ /dev/null
@@ -1,249 +0,0 @@
-#include <common.h>
-#include <command.h>
-#include <linux/delay.h>
-#include <linux/io.h>
-#include <linux/types.h>
-#include "pll.h"
-#include "uart.h"
-
-#ifdef DEV_IS_GOYA
-#define HL_PLL_NR		0x100
-#define HL_PLL_NF		0x104
-#define HL_PLL_OD		0x108
-#define HL_PLL_NB		0x10c
-#define HL_PLL_CFG		0x110
-#define HL_PLL_CFG_LOCK		(1 << 4)
-#define HL_PLL_LOCK_INTR	0x128
-#define HL_PLL_DATA_CHNG	0x130
-#define HL_PLL_RST		0x134
-#define HL_PLL_DIV_FACTOR	0x200
-#define HL_PLL_DIV_FACTOR_CMD	0x220
-#define HL_PLL_DIV_SEL		0x280
-#define HL_PLL_DIV_EN		0x2a0
-
-#define HL_PLL_MAX_WAIT_TIME	20	/* 20 us */
-#define HL_PLL_LOCK_TOUT	5000	/* 5 ms */
-
-#define USEC_PER_SEC		1000000
-#define CPU_CYCELS_PER_USEC	(1200000000 / USEC_PER_SEC)
-#define BYPASS_CYCLES_PER_USEC	(50000000 / USEC_PER_SEC)
-#define MIN_CYCLES_PER_ITER	5
-
-#define mmCPU_PLL_BASE		0x7FFC4A2000ull
-#define mmIC_PLL_BASE		0x7FFC4A3000ull
-#define mmPSOC_MME_PLL_BASE	0x7FFCC71000ull
-
-enum hl_pll_bypass_val {
-	HL_PLL_REF_CLK,
-	HL_PLL_PLL_CLK,
-	HL_PLL_DIV_REF,
-	HL_PLL_DIV_PLL,
-};
-
-enum hl_pll_div_out {
-	HL_PLL_DIV0,
-	HL_PLL_DIV1,
-	HL_PLL_DIV2,
-	HL_PLL_DIV3,
-	HL_PLL_DIV_SIZE,
-};
-
-struct hl_pll_cfg {
-	char	*name;
-	void	*base;
-	u32	nr;
-	u32	nf;
-	u32	od;
-	u32	nb;
-	u32	div_factor[HL_PLL_DIV_SIZE];
-};
-
-static struct hl_pll_cfg goya_pll_cfg[] = {
-	{.name = "CPU", .base = (void *)mmCPU_PLL_BASE,
-	 .nr = 0, .nf = 47, .od = 1, .nb = 23, .div_factor = {0, 23, 1, -1} },
-	{.name = "IC", .base = (void *)mmIC_PLL_BASE,
-	 .nr = 0, .nf = 63, .od = 15, .nb = 31, .div_factor = {0, 1, 1, 2} },
-	{.name = "MME", .base = (void *)mmPSOC_MME_PLL_BASE,
-	 .nr = 0, .nf = 63, .od = 15, .nb = 31, .div_factor = {0, 1, 2, 8} },
-};
-
-static int hl_pll_get_cfg(struct hl_pll_cfg **hl_pll_cfg)
-{
-	int pll_num = ARRAY_SIZE(goya_pll_cfg);
-
-	*hl_pll_cfg = goya_pll_cfg;
-
-	return pll_num;
-}
-
-static enum hl_pll_bypass_val hl_pll_bypass_get(struct hl_pll_cfg *cfg,
-							enum hl_pll_div_out out)
-{
-	int offset = out * sizeof(u32);
-
-	return readl(cfg->base + HL_PLL_DIV_SEL + offset);
-}
-
-static void hl_pll_udelay(unsigned int usec)
-{
-	long cnt;
-	/*
-	 * Assumptions:
-	 * 1) the CPU PLL offset in the array is 0
-	 * 2) DIV1 is the PLL output to timestamp counter.
-	 */
-	enum hl_pll_bypass_val bval =
-			hl_pll_bypass_get(goya_pll_cfg, HL_PLL_DIV1);
-
-	if (bval == HL_PLL_REF_CLK || bval == HL_PLL_DIV_REF)
-		cnt = BYPASS_CYCLES_PER_USEC;
-	else
-		cnt = CPU_CYCELS_PER_USEC;
-
-	cnt = (cnt * usec) / MIN_CYCLES_PER_ITER;
-
-	while (cnt--)
-		;
-}
-
-static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
-{
-	writel(reset, cfg->base + HL_PLL_RST);
-}
-
-static inline int hl_pll_is_locked(struct hl_pll_cfg *cfg)
-{
-	return readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
-}
-
-static inline void hl_pll_clear_lock_status(struct hl_pll_cfg *cfg)
-{
-	u32 reg = readl(cfg->base + HL_PLL_CFG);
-
-	writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
-}
-
-static int hl_pll_cfg(struct hl_pll_cfg *cfg)
-{
-	int tout = HL_PLL_LOCK_TOUT;
-
-	hl_pll_reset(cfg, true);
-
-	writel(cfg->nr, cfg->base + HL_PLL_NR);
-	writel(cfg->nf, cfg->base + HL_PLL_NF);
-	writel(cfg->od, cfg->base + HL_PLL_OD);
-	writel(cfg->nb, cfg->base + HL_PLL_NB);
-	writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
-
-	/* after updating the coefficients we need to wait >=5us */
-	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
-
-	hl_pll_reset(cfg, false);
-
-	/* Wait >500 (10us) reference cycles until PLL re-locks. */
-	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
-
-	/* to be on a safe side, poll on the PLL lock status bit */
-	while (!hl_pll_is_locked(cfg) && tout--)
-		hl_pll_udelay(1);
-
-	if (!tout) {
-		preboot_puts_raw("PLL: Timeout waiting for ");
-		preboot_puts_raw(cfg->name);
-		preboot_puts_raw(" PLL lock...\n\r");
-		return 1;
-	}
-
-	/* w/a the h/w erratum: lock status bit isn't getting cleared */
-	hl_pll_clear_lock_status(cfg);
-
-	return 0;
-}
-
-static void hl_pll_bypass(struct hl_pll_cfg *cfg, bool bypass)
-{
-	int sel;
-	void *reg;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		if (cfg->div_factor[sel] == -1)
-			continue;
-
-		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
-
-		if (bypass)
-			writel(HL_PLL_REF_CLK, reg);
-		else
-			writel(cfg->div_factor[sel] ?
-					HL_PLL_DIV_PLL : HL_PLL_PLL_CLK, reg);
-	}
-
-	/* just to be on a safe side, wait for bypass to propagate */
-	hl_pll_udelay(1);
-}
-
-static void hl_pll_div_init(struct hl_pll_cfg *cfg)
-{
-	int sel;
-	void *div_base;
-
-	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
-		if (cfg->div_factor[sel] == -1)
-			continue;
-
-		div_base = cfg->base + (sel * sizeof(u32));
-
-		writel(cfg->div_factor[sel], div_base + HL_PLL_DIV_FACTOR);
-		writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
-		writel(1, div_base + HL_PLL_DIV_EN);
-	}
-
-	/* just to be on a safe side, wait for DIV init to propagate */
-	hl_pll_udelay(1);
-}
-
-static void hl_pll_ack_int(struct hl_pll_cfg *cfg)
-{
-	writel(1, cfg->base + HL_PLL_LOCK_INTR);
-}
-
-static int hl_pll_set(struct hl_pll_cfg *cfg)
-{
-	int ret;
-
-	hl_pll_bypass(cfg, true);
-
-	ret = hl_pll_cfg(cfg);
-
-	hl_pll_div_init(cfg);
-
-	hl_pll_bypass(cfg, false);
-
-	hl_pll_ack_int(cfg);
-
-	return ret;
-}
-
-/**
- * hl_pll_init - Initialize PLL registers
- */
-void hl_pll_init(void)
-{
-	int i, pll_num;
-	struct hl_pll_cfg *hl_pll_cfg;
-	enum hl_pll_bypass_val bval;
-
-	pll_num = hl_pll_get_cfg(&hl_pll_cfg);
-
-	for (i = 0; i < pll_num; i++) {
-		/* DIV0 is the undivided PLL output */
-		bval = hl_pll_bypass_get(hl_pll_cfg + i, HL_PLL_DIV0);
-		if ((u64)hl_pll_cfg[i].base == mmCPU_PLL_BASE &&
-				bval != HL_PLL_REF_CLK &&
-				bval != HL_PLL_DIV_REF)
-			continue;
-
-		hl_pll_set(hl_pll_cfg + i);
-	}
-}
-#endif
diff --git a/pll.h b/pll.h
deleted file mode 100644
index b3cc113..0000000
--- a/pll.h
+++ /dev/null
@@ -1,18 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_PLL_H_
-#define _HL_PLL_H_
-
-
-#ifdef DEV_IS_GOYA
-void hl_pll_init(void);
-#else
-static inline void hl_pll_init(void) {}
-#endif
-
-#endif /* _HL_PLL_H_ */
diff --git a/preboot.c b/preboot.c
deleted file mode 100644
index ab32f63..0000000
--- a/preboot.c
+++ /dev/null
@@ -1,172 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <common.h>
-#include <linux/io.h>
-#include <mach/hl_regs.h>
-#include <mach/hl_fw_if.h>
-#include "auto_generated.h"
-#include <stddef.h>
-#include <stdbool.h>
-#include "preboot.h"
-#include "pll.h"
-#include "uart.h"
-#include "spi.h"
-#include "pci.h"
-
-#define BTL_COUNTER_MAX		BIT(22)
-#define VERSION_MAX_LEN		128
-
-#define PREBOOT_VER_STR	"Preboot version " PREBOOT_VER \
-	" (" PREBOOT_DATE " - " PREBOOT_TIME ")"
-
-/*
- * Using scratchpad registers for
- * passing flash layout to CPU SW
- */
-static void preboot_copy_prepare(unsigned int *spl_off)
-{
-	unsigned int spl_offset = 0xa0000;
-	unsigned int uboot_offset = 0xb0000;
-	unsigned int modifiable_image_pending;
-	unsigned int image_base_offset;
-	unsigned int next_image, next_image_stable, next_image_modifiable;
-	bool try_new_image = false;
-	char *spl_image = "candidate";
-
-	modifiable_image_pending = readl(mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
-	if (modifiable_image_pending == FLOPS_UPD_PENDING) {
-		next_image_modifiable =
-			(*(unsigned int *)(HL_SPI_BASE + 0x60004) & 0xff) - '1';
-		if (next_image_modifiable < 2) {
-			next_image = next_image_modifiable;
-			writel((next_image + FLOPS_UPD_NXT_IMG1),
-					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
-			try_new_image = true;
-		} else
-			writel(FLOPS_UPD_NOOP,
-					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
-	}
-
-	if (!try_new_image) {
-		spl_image = "stable";
-		next_image_stable =
-			(*(unsigned int *)(HL_SPI_BASE + 0x60000) & 0xff) - '1';
-		if (next_image_stable < 2)
-			next_image = next_image_stable;
-		else
-			next_image = 0;
-	}
-
-	if (!next_image)
-		image_base_offset = 0;
-	else
-		image_base_offset = 0x800000;
-
-	preboot_puts_raw("Preparing to run SPL ");
-	preboot_puts_raw(spl_image);
-	preboot_puts_raw(" (offset: 0x");
-	preboot_puth_raw(image_base_offset, false);
-	preboot_puts(")\n");
-
-	*spl_off = image_base_offset + spl_offset;
-	writel((image_base_offset + uboot_offset)
-			, mmPSOC_GLOBAL_CONF_SCRATCHPAD_30);
-}
-
-
-static void preboot_copy_spl(void)
-{
-	unsigned int  spl_offset;
-
-	preboot_copy_prepare(&spl_offset);
-
-	spi_copy_spl(spl_offset);
-}
-
-#ifdef DEV_IS_GOYA
-/* Handle BTL errate: exhaust timeout counter before any SPI access */
-static void preboot_delay(void)
-{
-	unsigned long cntpct, end;
-	int sysclk = SYS_REF_CLOCK, i;
-
-	/* timer init */
-	writel(sysclk, mmCPU_TIMESTAMP_CNTFID0);
-	writel(0x3, mmCPU_TIMESTAMP_CNTCR); /* EN | HDBG */
-
-	/* get begin counter value and calc end counter value */
-	asm volatile("mrs %0, cntpct_el0" : "=r" (cntpct));
-	end = cntpct + BTL_COUNTER_MAX;
-
-	for (i = 0; i < BTL_COUNTER_MAX; i++) {
-		asm volatile("mrs %0, cntpct_el0" : "=r" (cntpct));
-		if (cntpct >= end)
-			break;
-	}
-}
-
-#else
-#define preboot_delay()
-#endif
-
-#ifndef TARGET_IS_TESTER
-static const char *preboot_ver_str = PREBOOT_VER_STR;
-static char preboot_btl_ver_str[VERSION_MAX_LEN] = "BTL version ";
-
-static void preboot_version(void)
-{
-	u64 offset = (u64) preboot_btl_ver_str;
-	u32 btlver = readl(mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
-	int i, btl_ver_sz;
-
-	preboot_puts("\n");
-	preboot_puts_raw(preboot_btl_ver_str);
-	preboot_puth(btlver, true);
-	preboot_puts_raw(preboot_ver_str);
-	preboot_puts("\n");
-
-	/* get to the end of the predefined btl string */
-	for (i = 0; i < VERSION_MAX_LEN; i++)
-		if (preboot_btl_ver_str[i] == '\0')
-			break;
-
-	/* convert and copy the btl version */
-	i += preboot_uint2hexstring(&preboot_btl_ver_str[i], btlver, true);
-
-	/* separate the btl from preboot */
-	preboot_btl_ver_str[i++] = '\n';
-	btl_ver_sz = i;
-
-	/* copy the preboot version */
-	for (int j = 0; j < VERSION_MAX_LEN - btl_ver_sz - 1; j++) {
-		preboot_btl_ver_str[i++] = preboot_ver_str[j];
-		if (preboot_ver_str[j] == '\0')
-			break;
-	}
-
-	writel((u32)(offset & ~CONFIG_SYS_SDRAM_BASE),
-			mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
-}
-#else	/* TARGET_IS_TESTER */
-static inline void preboot_version(void) {}
-#endif	/* TARGET_IS_TESTER */
-
-void preboot_main(void)
-{
-	void (*spl_start)(void) = (void *)CONFIG_SPL_TEXT_BASE;
-
-	writel(CPU_BOOT_STATUS_IN_PREBOOT, mmPSOC_GLOBAL_CONF_WARM_REBOOT);
-
-	preboot_version();
-	preboot_pcie_init();
-	preboot_delay();
-	hl_pll_init();
-	preboot_copy_spl();
-
-	spl_start();
-}
diff --git a/preboot.h b/preboot.h
deleted file mode 100644
index 825c1d3..0000000
--- a/preboot.h
+++ /dev/null
@@ -1,13 +0,0 @@
-/*
- * Copyright (C) 2019 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_PREBOOT_H_
-#define _HL_PREBOOT_H_
-
-#define SYS_REF_CLOCK			50000000
-
-#endif /* _HL_PREBOOT_H_ */
diff --git a/preboot.lds b/preboot.lds
deleted file mode 100644
index f8e8474..0000000
--- a/preboot.lds
+++ /dev/null
@@ -1,24 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-MEMORY { .sram : ORIGIN = 0x7ff0000000,
-		LENGTH = 0x10000 }
-
-OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
-OUTPUT_ARCH(aarch64)
-ENTRY(_start)
-SECTIONS
-{
-	.text : {
-		. = ALIGN(8);
-		start.o (.text*)
-		preboot.o (.text*)
-		pll.o (.text*)
-	} >.sram
-
-	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
-}
diff --git a/spi.c b/spi.c
deleted file mode 100644
index 1315bd2..0000000
--- a/spi.c
+++ /dev/null
@@ -1,227 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <config.h>
-#include <linux/io.h>
-#include <linux/sizes.h>
-#include <linux/bitops.h>
-#include <mach/hl_regs.h>
-
-#include "uart.h"
-
-#define mmPSOC_SPI_CTRLR0		(HL_CFG_BASE + 0xC43000)
-#define mmPSOC_SPI_SSIENR		(HL_CFG_BASE + 0xC43008)
-#define mmPSOC_SPI_SER			(HL_CFG_BASE + 0xC43010)
-#define mmPSOC_SPI_BAUDR		(HL_CFG_BASE + 0xC43014)
-#define mmPSOC_SPI_TXFTLR		(HL_CFG_BASE + 0xC43018)
-#define mmPSOC_SPI_RXFTLR		(HL_CFG_BASE + 0xC4301C)
-#define mmPSOC_SPI_TXFLR		(HL_CFG_BASE + 0xC43020)
-#define mmPSOC_SPI_RXFLR		(HL_CFG_BASE + 0xC43024)
-#define mmPSOC_SPI_IMR			(HL_CFG_BASE + 0xC4302C)
-#define mmPSOC_SPI_DR0			(HL_CFG_BASE + 0xC43060)
-#define SPL_SIZE			SZ_64K
-#define CMD_WRITE_STATUS		0x01
-#define CMD_READ			0x03
-#define CMD_WRITE_DISABLE		0x04
-#define CMD_READ_STATUS			0x05
-#define CMD_WRITE_ENABLE		0x06
-#define CMD_READ_CONF			0x15
-#define READ_STATUS_WIP			BIT(0)
-#define READ_STATUS_BP0			BIT(2)
-#define READ_STATUS_BP1			BIT(3)
-#define READ_STATUS_BP2			BIT(4)
-#define READ_STATUS_BP3			BIT(5)
-#define READ_STATUS_SRWD		BIT(7)
-#define READ_STATUS_BP0_3		(READ_STATUS_BP0 | READ_STATUS_BP1 | \
-					READ_STATUS_BP2 | READ_STATUS_BP3)
-#define PROT_MASK			(READ_STATUS_BP0_3 | READ_STATUS_SRWD)
-#define frame_size_bits(b)		((b) << 16)
-
-/*
- * SSI device Chip Select (CS0) is muxed with GPIO53
- * Using this we can realy do SSI bus claim/release
- * GPIO53 is set through mmPSOC_GPIO3 where each
- * mmPSOC_GPIO serve 16 GPIOs
- * Note: CS0 is active low
- */
-static void spi_ssi_cs0_init(void)
-{
-	unsigned int reg, mask = BIT_MASK(53 % 16);
-
-	reg  = readl(mmPSOC_GPIO3_DATAOUT);
-	reg  &= ~mask;
-	writel(reg,  mmPSOC_GPIO3_DATAOUT);
-
-	reg = readl(mmPSOC_GPIO3_OUTENSET);
-	reg |=  mask;
-	writel(reg, mmPSOC_GPIO3_OUTENSET);
-}
-
-static void spi_claim_bus(void)
-{
-	writel(1, mmPSOC_GLOBAL_CONF_PAD_SEL_53);
-}
-
-static void spi_release_bus(void)
-{
-	writel(0, mmPSOC_GLOBAL_CONF_PAD_SEL_53);
-}
-
-static inline void spi_dev_update_frame_size(unsigned int len)
-{
-	writel(0, mmPSOC_SPI_SSIENR);
-	writel(frame_size_bits((len * 8) - 1), mmPSOC_SPI_CTRLR0);
-	writel(1, mmPSOC_SPI_SSIENR);
-}
-
-static inline int spi_fifo_is_empty(unsigned long level_reg)
-{
-	return !readl(level_reg);
-}
-
-static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
-{
-	unsigned int tmp;
-
-	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
-		;
-	writel(cmd, mmPSOC_SPI_DR0);
-
-	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
-		;
-	tmp = readl(mmPSOC_SPI_DR0);
-
-	if (val)
-		*val = tmp;
-}
-
-static void spi_special_cmd(unsigned int cmd, unsigned char *val)
-{
-	unsigned char tmp;
-
-	spi_dev_update_frame_size(sizeof(tmp));
-
-	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
-		;
-	writeb(cmd, mmPSOC_SPI_DR0);
-
-	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
-		;
-	tmp = readb(mmPSOC_SPI_DR0);
-
-	if (val)
-		*val = tmp;
-}
-
-/* Hold chip select (CS) during transaction */
-static void spi_char_cmd_csel(unsigned char cmd, unsigned char *val)
-{
-	spi_claim_bus();
-
-	spi_special_cmd(cmd, NULL);
-	if (val)
-		spi_special_cmd(0, val);
-
-	spi_release_bus();
-}
-
-static void spi_special_cmd_csel(unsigned int cmd, unsigned char *val,
-								size_t size)
-{
-	int i;
-
-	spi_claim_bus();
-
-	spi_dev_update_frame_size(sizeof(*val));
-
-	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
-		;
-	writeb(cmd, mmPSOC_SPI_DR0);
-	for (i = 0; i < size; i++) {
-		while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
-			;
-		writeb(val[i], mmPSOC_SPI_DR0);
-	}
-
-	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
-		;
-	readb(mmPSOC_SPI_DR0);
-	for (i = 0; i < size; i++) {
-		while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
-			;
-		readb(mmPSOC_SPI_DR0);
-	}
-
-	spi_release_bus();
-}
-
-static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
-{
-	unsigned char sr;
-
-	spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
-	spi_special_cmd_csel(cmd, buf, size);
-	do
-		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	while (sr & READ_STATUS_WIP);
-	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
-}
-
-static unsigned char spi_prot(void)
-{
-	unsigned char sr, cr, data[2];
-
-	spi_char_cmd_csel(CMD_READ_CONF, &cr);
-	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	sr |= PROT_MASK;
-	data[0] = sr;
-	data[1] = cr;
-	spi_common_write(CMD_WRITE_STATUS, data, sizeof(data));
-	return sr;
-}
-
-static void spi_read_spl(unsigned int offset)
-{
-	void *dst = (void *)CONFIG_SPL_TEXT_BASE;
-	unsigned int cmd = (CMD_READ << 24) | offset;
-	int i;
-
-	spi_claim_bus();
-	spi_dev_update_frame_size(sizeof(cmd));
-	spi_dev_dword_cmd(cmd, NULL);
-	for (i = 0; i < SPL_SIZE; i += sizeof(unsigned int))
-		spi_dev_dword_cmd(0, dst + i);
-	spi_dev_dword_cmd(0, NULL);
-	spi_release_bus();
-}
-
-static void spi_init(void)
-{
-	spi_ssi_cs0_init();
-	writel(0x00000000, mmPSOC_SPI_SSIENR);
-	writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
-	writel(0x00000000, mmPSOC_SPI_RXFTLR);
-	writel(0x00000000, mmPSOC_SPI_TXFTLR);
-	writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
-	writel(0x00000001, mmPSOC_SPI_SER);
-	writel(0x00000001, mmPSOC_SPI_SSIENR);
-}
-
-void spi_copy_spl(unsigned int spl_offset)
-{
-	unsigned char sr;
-
-	spi_init();
-	spi_read_spl(spl_offset);
-	sr = spi_prot();
-	preboot_puts("SPL copying from flash is completed\n");
-	if ((sr & PROT_MASK) == PROT_MASK)
-		preboot_puts("Flash may be protected, depend on WP# signal\n");
-	else
-		preboot_puts("Flash is not protected\n");
-
-}
diff --git a/spi.h b/spi.h
deleted file mode 100644
index 5d2051c..0000000
--- a/spi.h
+++ /dev/null
@@ -1,14 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_SPI_H_
-#define _HL_SPI_H_
-
-
-void spi_copy_spl(unsigned int spl_offset);
-
-#endif /* _HL_SPI_H_ */
diff --git a/start.S b/start.S
deleted file mode 100644
index f988923..0000000
--- a/start.S
+++ /dev/null
@@ -1,18 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-/*
- * This is preboot entry point
- * Setting sp/fp/lr and jumping to C code
- */
-.globl	_start
-_start:
-
-	ldr	x0, =(_start + 0x10000)
-	bic	sp, x0, #0xf	/* 16-byte alignment for ABI compliance */
-	mov	x29, sp		/* Set FP (x29) */
-	bl	preboot_main	/* Set LR (x30) */
diff --git a/uart.c b/uart.c
deleted file mode 100644
index 7177c76..0000000
--- a/uart.c
+++ /dev/null
@@ -1,128 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <linux/io.h>
-#include <mach/hl_regs.h>
-
-#define mmPSOC_UART_0_DATA		(HL_CFG_BASE + 0xC45000)
-#define mmPSOC_UART_0_STATUS		(HL_CFG_BASE + 0xC45004)
-#define UART_SR_TXFULL			(1 << 0) /* TX Buffer Full */
-#define STR_UINT_SZ			8
-
-#ifdef DEV_IS_GOYA
-static void preboot_putc_virt_uart(const char c)
-{
-	/*
-	 * update vUART to HOST
-	 * DATA: mmPCIE_DB_RSV_RSV_0
-	 * STATUS: mmPCIE_DB_RSV_RSV_1
-	 */
-	writel(c, mmPCIE_DB_RSV_RSV_0);
-	writel(1, mmPCIE_DB_RSV_RSV_1);
-}
-#else
-static inline void preboot_putc_virt_uart(const char c) {}
-#endif
-
-void preboot_puts_raw(const char *str)
-{
-	int i;
-
-	for (i = 0; str[i] && (i < 128); i++) {
-		while (readl(mmPSOC_UART_0_STATUS) & UART_SR_TXFULL)
-			;
-		writel(str[i], mmPSOC_UART_0_DATA);
-		preboot_putc_virt_uart(str[i]);
-	}
-}
-
-void preboot_puts(const char *str)
-{
-	preboot_puts_raw(str);
-	preboot_puts_raw("\r");
-}
-
-/*
- * preboot_byte2char	- convert half byte to char.
- *
- * This function takes a byte number and converts its value to
- * a character representing the value (e.g. 0x8c can be converted to
- * '8' or 'c').
- */
-static inline char preboot_byte2char(u8 byte, bool upper)
-{
-	/* upper or lower 4 bits of the byte */
-	byte = upper ? byte >> 4 : byte & 0xf;
-
-	/*
-	 * convert: if half byte is a number [0..9] => use '0' as an offset
-	 *	    if half byte is a letter [0xa..0xf] => use 'a' as a base,
-	 *	    but since 'a' is not 0, we need to subtract
-	 *	    its base offset - 0xa.
-	 */
-	return byte + ((byte >= 0xa) ? 'a' - 0xa : '0');
-}
-
-/*
- * preboot_uint2hexstring	- convert uint to a hex values string.
- *
- * Take a number and provide its hexadecimal string representation.
- */
-int preboot_uint2hexstring(char *str, unsigned int num, bool pad)
-{
-	int j = 0;
-	char tmp;
-
-	/*
-	 * Here we walk on each byte of uint num in the reverse order, so
-	 * we build the string in the correct order.
-	 */
-	for (int i = 3; i >= 0; i--) {
-		u8 byte = (u8)((num >> (i * 8)) & 0xff);
-		/* here we run the loop on 2 parts of byte: upper and lower */
-		for (int k = 0; k < 2; k++) {
-			tmp = preboot_byte2char(byte, !k);
-			if (tmp == '0')	/* care of the leading zeroes (pad) */
-				pad = true;
-			if (pad)
-				str[j++] = tmp;
-		}
-	}
-
-	/* for !pad, return how many characters were actually written */
-	return j;
-}
-
-void preboot_puth_raw(unsigned int val, bool pad)
-{
-	char str[STR_UINT_SZ + 1];
-	int val_sz;
-
-	val_sz = preboot_uint2hexstring(str, val, pad);
-	val_sz = val_sz > STR_UINT_SZ ? STR_UINT_SZ : val_sz;
-	str[val_sz] = '\0';
-	preboot_puts_raw(str);
-}
-
-void preboot_puth(unsigned int val, bool pad)
-{
-	preboot_puth_raw(val, pad);
-	preboot_puts("\n");
-}
-
-static inline void preboot_puts_hex(const char *log, unsigned int val, bool pad)
-{
-	preboot_puts_raw(log);
-	preboot_puts_raw("0x");
-	preboot_puth(val, pad);
-}
-
-static inline void preboot_puts_char(const char *str, unsigned char val)
-{
-	preboot_puts_hex(str, val, false);
-}
-
diff --git a/uart.h b/uart.h
deleted file mode 100644
index 0d298f6..0000000
--- a/uart.h
+++ /dev/null
@@ -1,18 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_UART_H_
-#define _HL_UART_H_
-
-
-void preboot_puts_raw(const char *str);
-void preboot_puts(const char *str);
-int preboot_uint2hexstring(char *str, unsigned int num, bool pad);
-void preboot_puth_raw(unsigned int val, bool pad);
-void preboot_puth(unsigned int val, bool pad);
-
-#endif /* _HL_UART_H_ */
